Crypto/PublicKey+Apple.mm

```diff
@@ -359,69 +379,197 @@ virtual int _sign(int/*mbedtls_md_type_t*/ mbedDigestAlgorithm,
+                              "Cert::save",
+                              "A certificate already exists with the same persistentID");
+            }
+            
+            for (Retained<Cert> cert = this; cert; cert = cert->next()) {
+                SecCertificateRef certRef = SecCertificateCreateWithData(kCFAllocatorDefault,
+                                                                         (CFDataRef)cert->data().copiedNSData());
+                if (!certRef)
+                    throwMbedTLSError(MBEDTLS_ERR_X509_INVALID_FORMAT);
+                CFAutorelease(certRef);
 
-            if (err == errSecDuplicateItem) {
-                // Keychain can only have one cert with the same label (common name).
-                // Delete the existing one, then retry:
-                CFStringRef commonName;
-                checkOSStatus( SecCertificateCopyCommonName(certRef, &commonName),
-                              "SecCertificateCopyCommonName",
-                              "Couldn't replace a certificate in the Keychain");
-                Log("...first removing existing certificate with label '%s'",
-                    ((__bridge NSString*)commonName).UTF8String);
-                NSDictionary* delParams = @ {
+                NSMutableDictionary* params = [NSMutableDictionary dictionaryWithDictionary: @{
                     (id)kSecClass:              (id)kSecClassCertificate,
-                    (id)kSecAttrLabel:          (__bridge id)commonName,
-                };
-                checkOSStatus( SecItemDelete((CFDictionaryRef)delParams),
-                              "SecItemDelete",
-                              "Couldn't replace a certificate in the Keychain");
-                // Now retry:
+                    (id)kSecValueRef:           (__bridge id)certRef,
+                    (id)kSecReturnRef:          @YES
+                }];
+                
+                if (cert == this) {
+                    // Set the label to the leaf cert:
+                    params[id(kSecAttrLabel)] = label;
+                }
+                
+                CFTypeRef result;
                 ++gC4ExpectExceptions;
-                err = SecItemAdd((CFDictionaryRef)params, &result);
+                OSStatus status = SecItemAdd((CFDictionaryRef)params, &result);
                 --gC4ExpectExceptions;
+                
+                if (result)
+                    CFAutorelease(result);
+                
+                if (status == errSecDuplicateItem && cert != this) {
+                    // Ignore duplicates as it might be referenced by the other certificates
+                    Log("Ignore adding the certificate to the Keychain for '%.*s' as duplicated",
+                        SPLAT(cert->subjectName()));
+                    continue;
+                } else
+                    checkOSStatus(status, "SecItemAdd", "Coulding add a certificate to the Keychain");
```

Comment:
typo?? `Couldn't add a certificate to the Keychain`

Reply:
Fixed.